### PR TITLE
Stop cleaning monitor updates on new block connect

### LINKED ISSUE #2706: Stop reading monitors when persisting in updating persister
Turns out this line is brutalizing our heap and leading to fragmentation, it needs to go away - https://github.com/lightningdevkit/rust-lightning/blob/281a0aead744351c43fce660cf78f955fc969bfb/lightning/src/util/persist.rs#L613

### PR DESCRIPTION
**Pull request by G8XSU**
Previously, we used to cleanup monitor updates at both consolidation threshold and new block connects. With this change we will only cleanup when our consolidation criteria is met. Also, we remove monitor read from cleanup logic, in case of update consolidation. 
Note: In case of channel-closing monitor update, we still need to read the old monitor before persisting the new one in order to determine the cleanup range.

Closes #2706

### CODE REVIEW

File: lightning/src/util/persist.rs [hunk 1]
```diff
@@ -761,6 +727,34 @@ where
...
 }
 
+impl<K: Deref, L: Deref, ES: Deref, SP: Deref> MonitorUpdatingPersister<K, L, ES, SP>
+where
+	ES::Target: EntropySource + Sized,
+	K::Target: KVStore,
+	L::Target: Logger,
+	SP::Target: SignerProvider + Sized
+{
+	// Cleans up monitor updates for given monitor in range `start..=end`.
```
Comment (domZippilli) [inline lightning/src/util/persist.rs:737]: triple slash for rustdoc?

```suggestion
	/// Cleans up monitor updates for given monitor in range `start..=end`.
```

**Comment by domZippilli** [comment]
I might be nitpicking, but I thought an expectation of the KVStore contract
is that the backend would do things like exponential back-off retries to
deal with throttling. So a failure from any `impl KVStore` would likely
indicate a serious problem.

File: lightning/src/util/persist.rs [hunk 2]
```diff
@@ -751,8 +682,43 @@ where
...
+							Some(Ok((_, ref old_monitor))) => {
+								let start = old_monitor.get_latest_update_id();
+								// We never persist an update with update_id = CLOSED_CHANNEL_UPDATE_ID
+								let end = cmp::min(
+									start.saturating_add(self.maximum_pending_updates),
+									CLOSED_CHANNEL_UPDATE_ID - 1,
+								);
+								Some((start, end))
+							}
+							_ => None
```
[the remark this answers is not in the document]
Comment (shaavan) [reply lightning/src/util/persist.rs]: I agree that panicking here, though potentially inconvenient for the user, can definitely save them from a big mishap later down the line.

However, I can also notice one thing. The `old_monitor` is only tried to be read once throughout the life of a channel that is when the channel is closed.

So, if IO errors are happening for a user, they should be detected by this time in the life of the channel. So a panic here, though potentially useful, will probably also be inconsequential in the grand scheme of things.

**Review by shaavan — COMMENTED** [review]
Post merge ACK

I also agree with not panicking here, as its potential usefulness is outweighed by the unnecessary preemptive failure it might create.

Other than this, the PR looks perfect! Thanks a lot for this great cleanup improvement!
<details>
<summary>
<b>Summary and notes:</b>
</summary>

1. Not needing to clean the monitor, in the case of `persist_new_channel` makes sense, as we are doing a full persist
2. Not needing to read the old_monitor, in case of `% maximum_pending_updates == 0`, to determine the cleanup range, is a great use of the fact of the deterministic length of cleanup.
3. Separating `cleanup_in_range` in a separate function makes the code modular and the functionality usable separately, a good preemptive measure for future updates. 
</details>

As a small follow-up, we can add a comment in `persist_new_channel` fn to explain why we don’t need to do an old_monitor read & clean up here.

**Pull request merged**